// @toruslabs/customauth/dist/lib.cjs/index.js
'use strict';

var HandlerFactory = require('./handlers/HandlerFactory.js');
var interfaces = require('./handlers/interfaces.js');
var login = require('./login.js');
var enums = require('./utils/enums.js');
var error = require('./utils/error.js');
var helpers = require('./utils/helpers.js');



exports.createHandler = HandlerFactory;
exports.EMAIL_FLOW = interfaces.EMAIL_FLOW;
exports.CustomAuth = login;
exports.AGGREGATE_VERIFIER = enums.AGGREGATE_VERIFIER;
exports.LOGIN = enums.LOGIN;
exports.REDIRECT_PARAMS_STORAGE_METHOD = enums.REDIRECT_PARAMS_STORAGE_METHOD;
exports.SENTRY_TXNS = enums.SENTRY_TXNS;
exports.TORUS_METHOD = enums.TORUS_METHOD;
exports.UX_MODE = enums.UX_MODE;
exports.serializeError = error.serializeError;
exports.are3PCSupported = helpers.are3PCSupported;
exports.broadcastChannelOptions = helpers.broadcastChannelOptions;
exports.constructURL = helpers.constructURL;
exports.decodeToken = helpers.decodeToken;
exports.eventToPromise = helpers.eventToPromise;
exports.getPopupFeatures = helpers.getPopupFeatures;
exports.getTimeout = helpers.getTimeout;
exports.getVerifierId = helpers.getVerifierId;
exports.handleRedirectParameters = helpers.handleRedirectParameters;
exports.isFirefox = helpers.isFirefox;
exports.isMobileOrTablet = helpers.isMobileOrTablet;
exports.loginToConnectionMap = helpers.loginToConnectionMap;
exports.padUrlString = helpers.padUrlString;
exports.randomId = helpers.randomId;
exports.storageAvailable = helpers.storageAvailable;
exports.validateAndConstructUrl = helpers.validateAndConstructUrl;
