// @tkey/service-provider-base/dist/lib.cjs/ServiceProviderBase.js
'use strict';

var _defineProperty = require('@babel/runtime/helpers/defineProperty');
var commonTypes = require('@tkey/common-types');
var BN = require('bn.js');

class ServiceProviderBase {
  constructor({
    enableLogging = false,
    postboxKey
  }) {
    _defineProperty(this, "enableLogging", void 0);
    // For easy serialization
    _defineProperty(this, "postboxKey", void 0);
    _defineProperty(this, "serviceProviderName", void 0);
    _defineProperty(this, "migratableKey", null);
    this.enableLogging = enableLogging;
    this.postboxKey = new BN(postboxKey, "hex");
    this.serviceProviderName = "ServiceProviderBase";
  }
  static fromJSON(value) {
    const {
      enableLogging,
      postboxKey,
      serviceProviderName
    } = value;
    if (serviceProviderName !== "ServiceProviderBase") return undefined;
    return new ServiceProviderBase({
      enableLogging,
      postboxKey
    });
  }
  async encrypt(msg) {
    const publicKey = this.retrievePubKey("ecc");
    return commonTypes.encrypt(publicKey, msg);
  }
  async decrypt(msg) {
    return commonTypes.decrypt(commonTypes.toPrivKeyECC(this.postboxKey), msg);
  }
  retrievePubKeyPoint() {
    return commonTypes.toPrivKeyEC(this.postboxKey).getPublic();
  }
  retrievePubKey(type) {
    if (type === "ecc") {
      return commonTypes.getPubKeyECC(this.postboxKey);
    }
    throw new Error("Unsupported pub key type");
  }
  sign(msg) {
    const tmp = new BN(msg, "hex");
    const sig = commonTypes.toPrivKeyEC(this.postboxKey).sign(tmp.toString("hex"));
    return Buffer.from(sig.r.toString(16, 64) + sig.s.toString(16, 64) + new BN(0).toString(16, 2), "hex").toString("base64");
  }
  toJSON() {
    return {
      enableLogging: this.enableLogging,
      postboxKey: this.postboxKey.toString("hex"),
      serviceProviderName: this.serviceProviderName
    };
  }
}
var ServiceProviderBase$1 = ServiceProviderBase;

module.exports = ServiceProviderBase$1;
