// @tkey/tss/dist/lib.cjs/index.js
'use strict';

var provider = require('./provider.js');
var tss = require('./tss.js');
var util = require('./util.js');
var rssClient = require('@toruslabs/rss-client');



exports.TSSTorusServiceProvider = provider.TSSTorusServiceProvider;
exports.FACTOR_KEY_TYPE = tss.FACTOR_KEY_TYPE;
exports.LEGACY_KEY_TYPE = tss.LEGACY_KEY_TYPE;
exports.TKeyTSS = tss.TKeyTSS;
exports.TSS_MODULE = tss.TSS_MODULE;
exports.TSS_TAG_DEFAULT = tss.TSS_TAG_DEFAULT;
exports.factorKeyCurve = tss.factorKeyCurve;
exports.DELIMITERS = util.DELIMITERS;
exports.generateSalt = util.generateSalt;
exports.getEd25519SeedStoreDomainKey = util.getEd25519SeedStoreDomainKey;
exports.getExtendedVerifierId = util.getExtendedVerifierId;
exports.getLagrangeCoeffs = util.getLagrangeCoeffs;
exports.getPubKeyPoint = util.getPubKeyPoint;
exports.kCombinations = util.kCombinations;
exports.lagrangeInterpolation = util.lagrangeInterpolation;
exports.pointToHex = util.pointToHex;
Object.defineProperty(exports, "randomSelection", {
	enumerable: true,
	get: function () { return rssClient.randomSelection; }
});
