// @tkey/common-types/dist/lib.esm/utils.js
import { serializeError } from '@toruslabs/customauth';
import { encrypt as encrypt$1, decrypt as decrypt$1, generatePrivate } from '@toruslabs/eccrypto';
import { keccak256, toChecksumAddress } from '@toruslabs/torus.js';
import BN from 'bn.js';
import { ec } from 'elliptic';

const secp256k1 = new ec("secp256k1");

// Wrappers around ECC encrypt/decrypt to use the hex serialization
// TODO: refactor to take BN
async function encrypt(publicKey, msg) {
  const encryptedDetails = await encrypt$1(publicKey, msg);
  return {
    ciphertext: encryptedDetails.ciphertext.toString("hex"),
    ephemPublicKey: encryptedDetails.ephemPublicKey.toString("hex"),
    iv: encryptedDetails.iv.toString("hex"),
    mac: encryptedDetails.mac.toString("hex")
  };
}
async function decrypt(privKey, msg) {
  const bufferEncDetails = {
    ciphertext: Buffer.from(msg.ciphertext, "hex"),
    ephemPublicKey: Buffer.from(msg.ephemPublicKey, "hex"),
    iv: Buffer.from(msg.iv, "hex"),
    mac: Buffer.from(msg.mac, "hex")
  };
  return decrypt$1(privKey, bufferEncDetails);
}
function isEmptyObject(obj) {
  return Object.keys(obj).length === 0 && obj.constructor === Object;
}
const isErrorObj = err => err && err.stack && err.message !== "";
async function prettyPrintError(error) {
  if (isErrorObj(error)) {
    return error;
  }
  return serializeError(error);
}
function generateAddressFromPublicKey(publicKey) {
  const ethAddressLower = `0x${keccak256(publicKey).slice(64 - 38)}`;
  return toChecksumAddress(ethAddressLower);
}
function normalize(input) {
  if (!input) {
    return undefined;
  }
  let hexString;
  if (typeof input === "number") {
    hexString = input.toString(16);
    if (hexString.length % 2) {
      hexString = `0${hexString}`;
    }
  }
  if (typeof input === "string") {
    hexString = input.toLowerCase();
  }
  return `0x${hexString}`;
}
function generatePrivateExcludingIndexes(shareIndexes) {
  const key = new BN(generatePrivate());
  if (shareIndexes.find(el => el.eq(key))) {
    return generatePrivateExcludingIndexes(shareIndexes);
  }
  return key;
}
const KEY_NOT_FOUND = "KEY_NOT_FOUND";
const SHARE_DELETED = "SHARE_DELETED";
function derivePubKeyXFromPolyID(polyID) {
  return polyID.split("|")[0].slice(2);
}
function stripHexPrefix(str) {
  if (str.slice(0, 2) === "0x") return str.slice(2);
  return str;
}
function generateID() {
  // Math.random should be unique because of its seeding algorithm.
  // Convert it to base 36 (numbers + letters), and grab the first 9 characters
  // after the decimal.
  return `${Math.random().toString(36).substr(2, 9)}`;
}

export { KEY_NOT_FOUND, SHARE_DELETED, decrypt, derivePubKeyXFromPolyID, encrypt, generateAddressFromPublicKey, generateID, generatePrivateExcludingIndexes, isEmptyObject, isErrorObj, normalize, prettyPrintError, secp256k1, stripHexPrefix };
