// @tkey/core/dist/lib.cjs/core.js
'use strict';

var _objectSpread = require('@babel/runtime/helpers/objectSpread2');
var _defineProperty = require('@babel/runtime/helpers/defineProperty');
var commonTypes = require('@tkey/common-types');
var torus_js = require('@toruslabs/torus.js');
var BN = require('bn.js');
var random = require('ethereum-cryptography/random');
var stringify = require('json-stable-stringify');
var authMetadata = require('./authMetadata.js');
var errors = require('./errors.js');
var lagrangeInterpolatePolynomial = require('./lagrangeInterpolatePolynomial.js');
var metadata = require('./metadata.js');

const ed25519SeedConst = "ed25519Seed";

// TODO: handle errors for get and set with retries

class ThresholdKey {
  constructor(args) {
    _defineProperty(this, "modules", void 0);
    _defineProperty(this, "enableLogging", void 0);
    _defineProperty(this, "serviceProvider", void 0);
    _defineProperty(this, "storageLayer", void 0);
    _defineProperty(this, "shares", void 0);
    _defineProperty(this, "lastFetchedCloudMetadata", void 0);
    _defineProperty(this, "metadata", void 0);
    _defineProperty(this, "manualSync", void 0);
    _defineProperty(this, "_localMetadataTransitions", void 0);
    _defineProperty(this, "_refreshMiddleware", void 0);
    _defineProperty(this, "_reconstructKeyMiddleware", void 0);
    _defineProperty(this, "_shareSerializationMiddleware", void 0);
    _defineProperty(this, "storeDeviceShare", void 0);
    _defineProperty(this, "haveWriteMetadataLock", void 0);
    _defineProperty(this, "serverTimeOffset", 0);
    // secp256k1 key
    _defineProperty(this, "privKey", void 0);
    _defineProperty(this, "_ed25519Seed", void 0);
    const {
      enableLogging = false,
      modules = {},
      serviceProvider,
      storageLayer,
      manualSync = false,
      serverTimeOffset
    } = args || {};
    this.enableLogging = enableLogging;
    this.serviceProvider = serviceProvider;
    this.storageLayer = storageLayer;
    this.modules = modules;
    this.shares = {};
    this.privKey = undefined;
    this.manualSync = manualSync;
    this._refreshMiddleware = {};
    this._reconstructKeyMiddleware = {};
    this._shareSerializationMiddleware = undefined;
    this.storeDeviceShare = undefined;
    this._localMetadataTransitions = [[], []];
    this.setModuleReferences(); // Providing ITKeyApi access to modules
    this.haveWriteMetadataLock = "";
    this.serverTimeOffset = serverTimeOffset;
  }
  get secp256k1Key() {
    if (typeof this.privKey !== "undefined") {
      return this.privKey;
    }
    return null;
  }
  get ed25519Key() {
    if (typeof this._ed25519Seed !== "undefined") {
      return this._ed25519Seed;
    }
    return null;
  }
  set secp256k1Key(privKey) {
    this.privKey = privKey;
  }
  set ed25519Key(seed) {
    this._ed25519Seed = seed;
  }
  static async fromJSON(value, args) {
    const {
      enableLogging,
      privKey,
      metadata: metadata$1,
      shares,
      _localMetadataTransitions,
      manualSync,
      lastFetchedCloudMetadata,
      serverTimeOffset
    } = value;
    const {
      storageLayer,
      serviceProvider,
      modules
    } = args;
    const tb = new ThresholdKey({
      enableLogging,
      storageLayer,
      serviceProvider,
      modules,
      manualSync,
      serverTimeOffset
    });
    if (privKey) tb.privKey = new BN(privKey, "hex");
    for (const key in shares) {
      if (Object.prototype.hasOwnProperty.call(shares, key)) {
        const shareStoreMapElement = shares[key];
        for (const shareElementKey in shareStoreMapElement) {
          if (Object.prototype.hasOwnProperty.call(shareStoreMapElement, shareElementKey)) {
            const shareStore = shareStoreMapElement[shareElementKey];
            shareStoreMapElement[shareElementKey] = commonTypes.ShareStore.fromJSON(shareStore);
          }
        }
      }
    }
    tb.shares = shares;

    // switch to deserialize local metadata transition based on Object.keys() of authMetadata, ShareStore's and, IMessageMetadata
    const AuthMetadataKeys = Object.keys(JSON.parse(stringify(new authMetadata(new metadata(new commonTypes.Point("0", "0")), new BN("0", "hex")))));
    const ShareStoreKeys = Object.keys(JSON.parse(stringify(new commonTypes.ShareStore(new commonTypes.Share("0", "0"), ""))));
    const sampleMessageMetadata = {
      message: "Sample message",
      dateAdded: Date.now()
    };
    const MessageMetadataKeys = Object.keys(sampleMessageMetadata);
    const localTransitionShares = [];
    const localTransitionData = [];
    _localMetadataTransitions[0].forEach((x, index) => {
      if (x) {
        localTransitionShares.push(new BN(x, "hex"));
      } else {
        localTransitionShares.push(undefined);
      }
      const keys = Object.keys(_localMetadataTransitions[1][index]);
      if (keys.length === AuthMetadataKeys.length && keys.every(val => AuthMetadataKeys.includes(val))) {
        const tempAuth = authMetadata.fromJSON(_localMetadataTransitions[1][index]);
        tempAuth.privKey = privKey;
        localTransitionData.push(tempAuth);
      } else if (keys.length === ShareStoreKeys.length && keys.every(val => ShareStoreKeys.includes(val))) {
        localTransitionData.push(commonTypes.ShareStore.fromJSON(_localMetadataTransitions[1][index]));
      } else if (keys.length === MessageMetadataKeys.length && keys.every(val => MessageMetadataKeys.includes(val))) {
        localTransitionData.push(_localMetadataTransitions[1][index]);
      } else {
        throw errors.default("fromJSON failed. Could not deserialise _localMetadataTransitions");
      }
    });
    if (metadata$1 || lastFetchedCloudMetadata) {
      let tempMetadata;
      let tempCloud;
      let shareToUseForSerialization;

      // if service provider key is missing, we should initialize with one of the existing shares
      // TODO: fix for deleted share
      if (tb.serviceProvider.postboxKey.toString("hex") === "0") {
        const latestPolyIDOnCloud = metadata.fromJSON(lastFetchedCloudMetadata).getLatestPublicPolynomial().getPolynomialID();
        const shareIndexesExistInSDK = Object.keys(shares[latestPolyIDOnCloud]);
        const randomIndex = shareIndexesExistInSDK[Math.floor(Math.random() * (shareIndexesExistInSDK.length - 1))];
        if (shareIndexesExistInSDK.length >= 1) {
          shareToUseForSerialization = shares[latestPolyIDOnCloud][randomIndex];
        }
      }
      if (metadata$1) tempMetadata = metadata.fromJSON(metadata$1);
      if (lastFetchedCloudMetadata) tempCloud = metadata.fromJSON(lastFetchedCloudMetadata);
      await tb.initialize({
        neverInitializeNewKey: true,
        transitionMetadata: tempMetadata,
        previouslyFetchedCloudMetadata: tempCloud,
        previousLocalMetadataTransitions: [localTransitionShares, localTransitionData],
        withShare: shareToUseForSerialization
      });
    } else {
      await tb.initialize({
        neverInitializeNewKey: true
      });
    }
    return tb;
  }
  getStorageLayer() {
    return this.storageLayer;
  }
  getMetadata() {
    if (typeof this.metadata !== "undefined") {
      return this.metadata;
    }
    throw errors.metadataUndefined();
  }
  async initialize(params) {
    // setup initial params/states
    const p = params || {};
    if (p.delete1OutOf1 && !this.manualSync) throw errors.delete1OutOf1OnlyManualSync();
    const {
      withShare,
      importKey,
      importEd25519Seed,
      neverInitializeNewKey,
      transitionMetadata,
      previouslyFetchedCloudMetadata,
      previousLocalMetadataTransitions
    } = p;
    const previousLocalMetadataTransitionsExists = previousLocalMetadataTransitions && previousLocalMetadataTransitions[0].length > 0 && previousLocalMetadataTransitions[1].length > 0;
    const reinitializing = transitionMetadata && previousLocalMetadataTransitionsExists; // are we reinitializing the SDK?
    // in the case we're reinitializing whilst newKeyAssign has not been synced
    const reinitializingWithNewKeyAssign = reinitializing && previouslyFetchedCloudMetadata === undefined;
    let shareStore;
    if (withShare instanceof commonTypes.ShareStore) {
      shareStore = withShare;
    } else if (typeof withShare === "object") {
      shareStore = commonTypes.ShareStore.fromJSON(withShare);
    } else if (!withShare) {
      // default to use service provider
      // first we see if a share has been kept for us
      const spIncludeLocalMetadataTransitions = reinitializingWithNewKeyAssign;
      const spLocalMetadataTransitions = reinitializingWithNewKeyAssign ? previousLocalMetadataTransitions : undefined;
      const rawServiceProviderShare = await this.getGenericMetadataWithTransitionStates({
        serviceProvider: this.serviceProvider,
        includeLocalMetadataTransitions: spIncludeLocalMetadataTransitions,
        _localMetadataTransitions: spLocalMetadataTransitions,
        fromJSONConstructor: {
          fromJSON(val) {
            return val;
          }
        }
      });
      const noKeyFound = rawServiceProviderShare;
      if (noKeyFound.message === commonTypes.KEY_NOT_FOUND) {
        if (neverInitializeNewKey) {
          throw errors.default("key has not been generated yet");
        }

        // no metadata set, assumes new user
        // check for serviceprovider migratableKey for import key from service provider for new user
        // provided no importKey is provided ( importKey take precedent )
        if (this.serviceProvider.migratableKey && !(importKey || importEd25519Seed)) {
          // importkey from server provider need to be atomic, hence manual sync is required.
          const tempStateManualSync = this.manualSync; // temp store manual sync flag
          this.manualSync = true; // Setting this as true since _initializeNewKey has a check where for importkey from server provider need to be atomic, hence manual sync is required.
          await this._initializeNewKey({
            initializeModules: true,
            importedKey: this.serviceProvider.migratableKey,
            delete1OutOf1: true
          });
          if (!tempStateManualSync) await this.syncLocalMetadataTransitions(); // Only sync if we were not in manual sync mode, if manual sync is set by developer, they should handle it themselves
          // restore manual sync flag
          this.manualSync = tempStateManualSync;
        } else {
          await this._initializeNewKey({
            initializeModules: true,
            importedKey: importKey,
            delete1OutOf1: p.delete1OutOf1,
            importEd25519Seed
          });
        }

        // return after created new tkey account ( skip other steps)
        return this.getKeyDetails();
      }
      // else we continue with catching up share and metadata
      shareStore = commonTypes.ShareStore.fromJSON(rawServiceProviderShare);
    } else {
      throw errors.default("Input is not supported");
    }

    // We determine the latest metadata on the SDK and if there has been
    // needed transitions to include
    let currentMetadata;
    let latestCloudMetadata;
    // we fetch the latest metadata for the account from the share
    let latestShareDetails;
    try {
      latestShareDetails = await this.catchupToLatestShare({
        shareStore
      });
    } catch (error) {
      // check if error is not the undefined error
      // if so we don't throw immediately incase there is valid transition metadata
      const err = error;
      const noMetadataExistsForShare = err.code === 1503;
      if (!noMetadataExistsForShare || !reinitializing) {
        throw err;
      }
    }

    // lets check if the cloud metadata has been updated or not from previously if we are reinitializing
    if (reinitializing && !reinitializingWithNewKeyAssign) {
      if (previouslyFetchedCloudMetadata.nonce < latestShareDetails.shareMetadata.nonce) {
        throw errors.fromCode(1104);
      } else if (previouslyFetchedCloudMetadata.nonce > latestShareDetails.shareMetadata.nonce) {
        throw errors.fromCode(1105);
      }
      latestCloudMetadata = previouslyFetchedCloudMetadata;
    } else {
      latestCloudMetadata = latestShareDetails ? latestShareDetails.shareMetadata.clone() : undefined;
    }

    // If we've been provided with transition metadata we use that as the current metadata instead
    // as we want to maintain state before and after serialization.
    // (Given that the checks for cloud metadata pass)
    if (reinitializing) {
      currentMetadata = transitionMetadata;
      this._localMetadataTransitions = previousLocalMetadataTransitions;
    } else {
      currentMetadata = latestShareDetails.shareMetadata;
    }
    this.lastFetchedCloudMetadata = latestCloudMetadata;
    this.metadata = currentMetadata;
    const latestShare = latestShareDetails ? latestShareDetails.latestShare : shareStore;
    this.inputShareStore(latestShare);
    if (importEd25519Seed && this.getEd25519PublicKey()) {
      throw errors.default("Ed25119 key already exists");
    }

    // initialize modules
    await this.initializeModules();
    return this.getKeyDetails();
  }

  /**
   * catchupToLatestShare recursively loops fetches metadata of the provided share and checks if there is an encrypted share for it.
   * @param shareStore - share to start of with
   * @param polyID - if specified, polyID to refresh to if it exists
   */
  async catchupToLatestShare(params) {
    const {
      shareStore,
      polyID,
      includeLocalMetadataTransitions
    } = params;
    let shareMetadata;
    try {
      shareMetadata = await this.getAuthMetadata({
        privKey: shareStore.share.share,
        includeLocalMetadataTransitions
      });
    } catch (error) {
      // delete share error
      const err = error;
      if (err && err.code === 1308) {
        throw err;
      }
      const prettyError = await commonTypes.prettyPrintError(err);
      throw errors.authMetadataGetUnavailable(`, ${prettyError.message}`);
    }
    try {
      // if matches specified polyID return it
      if (polyID) {
        if (shareStore.polynomialID === polyID) {
          return {
            latestShare: shareStore,
            shareMetadata
          };
        }
      }
      const nextShare = await shareMetadata.getEncryptedShare(shareStore);
      return await this.catchupToLatestShare({
        shareStore: nextShare,
        polyID,
        includeLocalMetadataTransitions
      });
    } catch (error) {
      // delete share error
      const err = error;
      if (err && err.code === 1308) {
        throw err;
      }
      return {
        latestShare: shareStore,
        shareMetadata
      };
    }
  }
  async reconstructKey(_reconstructKeyMiddleware = true) {
    if (!this.metadata) {
      throw errors.metadataUndefined();
    }
    const pubPoly = this.metadata.getLatestPublicPolynomial();
    const requiredThreshold = pubPoly.getThreshold();
    const pubPolyID = pubPoly.getPolynomialID();

    // check if we have enough shares to meet threshold
    let sharesLeft = requiredThreshold;
    // we don't just check the latest poly but
    // we check if the shares on previous polynomials in our stores have the share indexes we require
    const fullShareList = this.metadata.getShareIndexesForPolynomial(pubPolyID);
    const shareIndexesRequired = {};
    for (let i = 0; i < fullShareList.length; i += 1) {
      shareIndexesRequired[fullShareList[i]] = true;
    }
    const sharesToInput = [];
    for (let z = this.metadata.polyIDList.length - 1; z >= 0 && sharesLeft > 0; z -= 1) {
      const sharesForPoly = this.shares[this.metadata.polyIDList[z][0]];
      if (sharesForPoly) {
        const shareIndexesForPoly = Object.keys(sharesForPoly);
        for (let k = 0; k < shareIndexesForPoly.length && sharesLeft > 0; k += 1) {
          if (shareIndexesForPoly[k] in shareIndexesRequired) {
            const currentShareForPoly = sharesForPoly[shareIndexesForPoly[k]];
            if (currentShareForPoly.polynomialID === pubPolyID) {
              sharesToInput.push(currentShareForPoly);
            } else {
              const latestShareRes = await this.catchupToLatestShare({
                shareStore: currentShareForPoly,
                polyID: pubPolyID,
                includeLocalMetadataTransitions: true
              });
              if (latestShareRes.latestShare.polynomialID === pubPolyID) {
                sharesToInput.push(latestShareRes.latestShare);
              } else {
                throw new errors(1304, "Share found in unexpected polynomial"); // Share found in unexpected polynomial
              }
            }
            delete shareIndexesRequired[shareIndexesForPoly[k]];
            sharesLeft -= 1;
          }
        }
      }
    }

    // Input shares to ensure atomicity
    sharesToInput.forEach(share => {
      this.inputShareStore(share);
    });
    if (sharesLeft > 0) {
      throw errors.unableToReconstruct(` require ${requiredThreshold} but have ${requiredThreshold - sharesLeft}`);
    }
    const polyShares = Object.keys(this.shares[pubPolyID]);
    const shareArr = [];
    const shareIndexArr = [];
    for (let i = 0; i < requiredThreshold; i += 1) {
      shareArr.push(this.shares[pubPolyID][polyShares[i]].share.share);
      shareIndexArr.push(this.shares[pubPolyID][polyShares[i]].share.shareIndex);
    }
    const privKey = lagrangeInterpolatePolynomial.lagrangeInterpolation(shareArr, shareIndexArr);
    // check that priv key regenerated is correct
    const reconstructedPubKey = commonTypes.getPubKeyPoint(privKey);
    if (this.metadata.pubKey.x.cmp(reconstructedPubKey.x) !== 0) {
      throw errors.incorrectReconstruction();
    }
    this.secp256k1Key = privKey;
    const returnObject = {
      allKeys: [privKey]
    };
    if (_reconstructKeyMiddleware && Object.keys(this._reconstructKeyMiddleware).length > 0) {
      // retireve/reconstruct extra keys that live on metadata
      await Promise.all(Object.keys(this._reconstructKeyMiddleware).map(async x => {
        if (Object.prototype.hasOwnProperty.call(this._reconstructKeyMiddleware, x)) {
          const extraKeys = await this._reconstructKeyMiddleware[x]();
          returnObject[x] = extraKeys;
          returnObject.allKeys.push(...extraKeys);
        }
      }));
    }

    // ed25519key
    if (this.getEd25519PublicKey()) {
      const seed = await this.retrieveEd25519Seed();
      if (!seed) {
        throw errors.default("Ed25519 seed not found");
      }
      this._ed25519Seed = seed;
    }
    return _objectSpread({
      secp256k1Key: privKey,
      ed25519Seed: this._ed25519Seed
    }, returnObject);
  }
  reconstructLatestPoly() {
    if (!this.metadata) {
      throw errors.metadataUndefined();
    }
    const pubPoly = this.metadata.getLatestPublicPolynomial();
    const pubPolyID = pubPoly.getPolynomialID();
    const threshold = pubPoly.getThreshold();
    const pointsArr = [];
    const sharesForExistingPoly = Object.keys(this.shares[pubPolyID]);
    if (sharesForExistingPoly.length < threshold) {
      throw errors.unableToReconstruct("not enough shares to reconstruct poly");
    }
    if (new Set(sharesForExistingPoly).size !== sharesForExistingPoly.length) {
      throw errors.default("share indexes should be unique");
    }
    for (let i = 0; i < threshold; i += 1) {
      pointsArr.push(new commonTypes.Point(new BN(sharesForExistingPoly[i], "hex"), this.shares[pubPolyID][sharesForExistingPoly[i]].share.share));
    }
    return lagrangeInterpolatePolynomial.lagrangeInterpolatePolynomial(pointsArr);
  }
  async deleteShare(shareIndex) {
    if (!this.metadata) {
      throw errors.metadataUndefined();
    }
    if (!this.privKey) {
      throw errors.privateKeyUnavailable();
    }
    const shareIndexToDelete = new BN(shareIndex, "hex");
    const shareToDelete = this.outputShareStore(shareIndexToDelete);
    if (shareIndexToDelete.cmp(new BN("1", "hex")) === 0) {
      throw new errors(1001, "Unable to delete service provider share");
    }

    // Get existing shares
    const pubPoly = this.metadata.getLatestPublicPolynomial();
    const previousPolyID = pubPoly.getPolynomialID();
    const existingShareIndexes = this.metadata.getShareIndexesForPolynomial(previousPolyID);
    const newShareIndexes = [];
    existingShareIndexes.forEach(el => {
      const bn = new BN(el, "hex");
      if (bn.cmp(shareIndexToDelete) !== 0) {
        newShareIndexes.push(bn.toString("hex"));
      }
    });

    // Update shares
    if (existingShareIndexes.length === newShareIndexes.length) {
      throw errors.default("Share index does not exist in latest polynomial");
    } else if (newShareIndexes.length < pubPoly.getThreshold()) {
      throw errors.default(`Minimum ${pubPoly.getThreshold()} shares are required for tkey. Unable to delete share`);
    }
    const results = await this._refreshShares(pubPoly.getThreshold(), [...newShareIndexes], previousPolyID);
    const newShareStores = results.shareStores;
    await this.addLocalMetadataTransitions({
      input: [{
        message: commonTypes.SHARE_DELETED,
        dateAdded: Date.now()
      }],
      privKey: [shareToDelete.share.share]
    });
    return {
      newShareStores
    };
  }
  async generateNewShare() {
    if (!this.metadata) {
      throw errors.metadataUndefined();
    }
    if (!this.secp256k1Key) {
      throw errors.privateKeyUnavailable();
    }
    const pubPoly = this.metadata.getLatestPublicPolynomial();
    const previousPolyID = pubPoly.getPolynomialID();
    const existingShareIndexes = this.metadata.getShareIndexesForPolynomial(previousPolyID);
    const existingShareIndexesBN = existingShareIndexes.map(el => new BN(el, "hex"));
    const newShareIndex = new BN(commonTypes.generatePrivateExcludingIndexes(existingShareIndexesBN));
    const results = await this._refreshShares(pubPoly.getThreshold(), [...existingShareIndexes, newShareIndex.toString("hex")], previousPolyID);
    const newShareStores = results.shareStores;
    return {
      newShareStores,
      newShareIndex
    };
  }
  getEd25519PublicKey() {
    if (!this.metadata) {
      throw errors.metadataUndefined();
    }
    const result = this.metadata.getGeneralStoreDomain(ed25519SeedConst);
    return result === null || result === void 0 ? void 0 : result.publicKey;
  }
  async retrieveEd25519Seed() {
    if (!this.metadata) {
      throw errors.metadataUndefined();
    }
    if (!this.privKey) {
      throw errors.privateKeyUnavailable();
    }
    const result = this.metadata.getGeneralStoreDomain(ed25519SeedConst);
    const seed = await this.decrypt(result.message);
    this._ed25519Seed = seed;
    return seed;
  }
  async addLocalMetadataTransitions(params) {
    const {
      privKey,
      input
    } = params;
    this._localMetadataTransitions[0] = [...this._localMetadataTransitions[0], ...privKey];
    this._localMetadataTransitions[1] = [...this._localMetadataTransitions[1], ...input];
    if (!this.manualSync) await this.syncLocalMetadataTransitions();
  }
  async syncLocalMetadataTransitions() {
    if (!this.metadata) {
      throw errors.metadataUndefined();
    }
    if (!(Array.isArray(this._localMetadataTransitions[0]) && this._localMetadataTransitions[0].length > 0)) return;

    // get lock
    let acquiredLock = false;
    if (this.lastFetchedCloudMetadata) {
      await this.acquireWriteMetadataLock();
      acquiredLock = true;
    }
    try {
      await this.storageLayer.setMetadataStream({
        input: this._localMetadataTransitions[1],
        privKey: this._localMetadataTransitions[0],
        serviceProvider: this.serviceProvider
      });
      this._localMetadataTransitions = [[], []];
      this.lastFetchedCloudMetadata = this.metadata.clone();
    } catch (error) {
      const prettyError = await commonTypes.prettyPrintError(error);
      throw errors.metadataPostFailed(prettyError.message);
    } finally {
      // release lock
      if (acquiredLock) await this.releaseWriteMetadataLock();
    }
  }
  async readMetadata(privKey) {
    return this.storageLayer.getMetadata({
      privKey
    });
  }

  // Returns a new instance of metadata with a clean state. All the previous state will be reset.
  async updateSDK(params) {
    const tb = new ThresholdKey({
      enableLogging: this.enableLogging,
      modules: this.modules,
      serviceProvider: this.serviceProvider,
      storageLayer: this.storageLayer,
      manualSync: this.manualSync
    });
    try {
      await tb.initialize({
        neverInitializeNewKey: true,
        withShare: params && params.withShare
      });
    } catch (err) {
      throw errors.fromCode(1103, `${err.message}`);
    }

    // Delete unnecessary polyIDs and shareStores
    const allPolyIDList = tb.metadata.polyIDList;
    let lastValidPolyID;
    Object.keys(this.shares).forEach(x => {
      if (allPolyIDList.find(id => id[0] === x)) {
        lastValidPolyID = x;
      } else {
        delete this.shares[x];
      }
    });

    // catchup to latest shareStore for all latest available shares.
    // TODO: fix edge cases where shares are deleted in the newer polynomials
    // TODO: maybe assign this.shares directly rather than output and inputsharestore.
    const shareStoresForLastValidPolyID = Object.keys(this.shares[lastValidPolyID]).map(x => tb.inputShareStoreSafe(this.outputShareStore(x, lastValidPolyID)));
    await Promise.all(shareStoresForLastValidPolyID);
    return tb;
  }

  // NOTE: This API will be DEPRECATED in the future in favour of inputShareStoreSafe()
  inputShareStore(shareStore) {
    let ss;
    if (shareStore instanceof commonTypes.ShareStore) {
      ss = shareStore;
    } else if (typeof shareStore === "object") {
      ss = commonTypes.ShareStore.fromJSON(shareStore);
    } else {
      throw errors.default("can only add type ShareStore into shares");
    }
    if (!(ss.polynomialID in this.shares)) {
      this.shares[ss.polynomialID] = {};
    }
    this.shares[ss.polynomialID][ss.share.shareIndex.toString("hex")] = ss;
  }

  // inputs a share ensuring that the share is the latest share AND metadata is updated to its latest state
  async inputShareStoreSafe(shareStore, autoUpdateMetadata = false) {
    if (!this.metadata) {
      throw errors.metadataUndefined();
    }
    let ss;
    if (shareStore instanceof commonTypes.ShareStore) {
      ss = shareStore;
    } else if (typeof shareStore === "object") {
      ss = commonTypes.ShareStore.fromJSON(shareStore);
    } else {
      throw errors.default("can only add type ShareStore into shares");
    }
    const polynomialId = this.metadata.getLatestPublicPolynomial().getPolynomialID();
    if (ss.polynomialID !== polynomialId) {
      const latestShareRes = await this.catchupToLatestShare({
        shareStore: ss,
        includeLocalMetadataTransitions: true
      });
      // check if the latest share is part of the current tkey instances
      // to avoid random share getting input into metadata
      if (!latestShareRes.shareMetadata.polyIDList.find(tuple => tuple[0] === polynomialId)) {
        throw errors.fromCode(1307);
      }
      // if latest share's polynomial is not equal with tkey latest polynomial, tkey's metadata is outdated
      if (polynomialId !== latestShareRes.latestShare.polynomialID) {
        if (!autoUpdateMetadata) throw errors.default(`TKey SDK metadata seems to be outdated because shareIndex: ` + `${latestShareRes.latestShare.share.shareIndex.toString("hex")} has a more recent metadata. Please call updateSDK first`);else this.metadata = latestShareRes.shareMetadata;
      }
      if (!(latestShareRes.latestShare.polynomialID in this.shares)) {
        this.shares[latestShareRes.latestShare.polynomialID] = {};
      }
      this.shares[latestShareRes.latestShare.polynomialID][latestShareRes.latestShare.share.shareIndex.toString("hex")] = latestShareRes.latestShare;
    } else {
      if (!(ss.polynomialID in this.shares)) {
        this.shares[ss.polynomialID] = {};
      }
      this.shares[ss.polynomialID][ss.share.shareIndex.toString("hex")] = ss;
    }
  }
  outputShareStore(shareIndex, polyID) {
    if (!this.metadata) {
      throw errors.metadataUndefined();
    }
    let shareIndexParsed;
    if (typeof shareIndex === "number") {
      shareIndexParsed = new BN(shareIndex);
    } else if (BN.isBN(shareIndex)) {
      shareIndexParsed = shareIndex;
    } else if (typeof shareIndex === "string") {
      shareIndexParsed = new BN(shareIndex, "hex");
    }
    let polyIDToSearch;
    if (polyID) {
      polyIDToSearch = polyID;
    } else {
      polyIDToSearch = this.metadata.getLatestPublicPolynomial().getPolynomialID();
    }
    if (!this.metadata.getShareIndexesForPolynomial(polyIDToSearch).includes(shareIndexParsed.toString("hex"))) {
      throw new errors(1002, "no such share index created");
    }
    const shareFromStore = this.shares[polyIDToSearch][shareIndexParsed.toString("hex")];
    if (shareFromStore) return shareFromStore;
    const poly = this.reconstructLatestPoly();
    const shareMap = poly.generateShares([shareIndexParsed]);
    return new commonTypes.ShareStore(shareMap[shareIndexParsed.toString("hex")], polyIDToSearch);
  }
  getCurrentShareIndexes() {
    if (!this.metadata) {
      throw errors.metadataUndefined();
    }
    const latestPolynomial = this.metadata.getLatestPublicPolynomial();
    const latestPolynomialId = latestPolynomial.getPolynomialID();
    const currentShareIndexes = Object.keys(this.shares[latestPolynomialId]);
    return currentShareIndexes;
  }
  getKeyDetails() {
    if (!this.metadata) {
      throw errors.metadataUndefined();
    }
    const poly = this.metadata.getLatestPublicPolynomial();
    const previousPolyID = poly.getPolynomialID();
    const requiredShares = poly.getThreshold() - Object.keys(this.shares[previousPolyID]).length;
    let shareDescriptions = this.metadata.getShareDescription();
    if (shareDescriptions) {
      const existingShareIndexes = this.metadata.getShareIndexesForPolynomial(previousPolyID);
      shareDescriptions = Object.keys(shareDescriptions).reduce((acc, index) => {
        if (existingShareIndexes.indexOf(index) >= 0) acc[index] = shareDescriptions[index];
        return acc;
      }, {});
    }
    return {
      pubKey: this.metadata.pubKey,
      ed25519PublicKey: this.getEd25519PublicKey(),
      requiredShares,
      threshold: poly.getThreshold(),
      totalShares: this.metadata.getShareIndexesForPolynomial(previousPolyID).length,
      shareDescriptions
    };
  }

  // Auth functions

  generateAuthMetadata(params) {
    const {
      input
    } = params;
    const authMetadatas = [];
    for (let i = 0; i < input.length; i += 1) {
      authMetadatas.push(new authMetadata(input[i], this.privKey));
    }
    return authMetadatas;
  }
  setAuthMetadata(params) {
    const {
      input,
      serviceProvider,
      privKey
    } = params;
    const authMetadata$1 = new authMetadata(input, this.privKey);
    return this.storageLayer.setMetadata({
      input: authMetadata$1,
      serviceProvider,
      privKey
    });
  }
  async setAuthMetadataBulk(params) {
    if (!this.privKey) {
      throw errors.privateKeyUnavailable();
    }
    const {
      input,
      serviceProvider,
      privKey
    } = params;
    const authMetadatas = [];
    for (let i = 0; i < input.length; i += 1) {
      authMetadatas.push(new authMetadata(input[i], this.privKey));
    }
    await this.addLocalMetadataTransitions({
      input: authMetadatas,
      serviceProvider,
      privKey
    });
  }
  async getAuthMetadata(params) {
    const raw = await this.getGenericMetadataWithTransitionStates(_objectSpread(_objectSpread({}, params), {}, {
      fromJSONConstructor: authMetadata
    }));
    const authMetadata$1 = raw;
    return authMetadata$1.metadata;
  }

  // fetches the latest metadata potentially searching in local transition states first
  async getGenericMetadataWithTransitionStates(params) {
    if (!(params.serviceProvider && params.serviceProvider.postboxKey.toString("hex") !== "0" || params.privKey)) {
      throw errors.default("require either serviceProvider or priv key in getGenericMetadataWithTransitionStates");
    }
    if (params.includeLocalMetadataTransitions) {
      const transitions = params._localMetadataTransitions ? params._localMetadataTransitions : this._localMetadataTransitions;
      let index = null;
      for (let i = transitions[0].length - 1; i >= 0; i -= 1) {
        const x = transitions[0][i];
        if (params.privKey && x && x.cmp(params.privKey) === 0) index = i;else if (params.serviceProvider && !x) index = i;
        if (index !== null) break;
      }
      if (index !== null) {
        return transitions[1][index];
      }
    }
    let raw;
    try {
      raw = await this.storageLayer.getMetadata(params);
    } catch (err) {
      const prettyError = await commonTypes.prettyPrintError(err);
      throw errors.metadataGetFailed(prettyError.message);
    }
    if (raw.message === commonTypes.SHARE_DELETED) {
      throw errors.fromCode(1308);
    }
    return params.fromJSONConstructor.fromJSON(raw);
  }

  // Lock functions
  async acquireWriteMetadataLock() {
    if (this.haveWriteMetadataLock) return this.metadata.nonce;
    if (!this.privKey) {
      throw errors.privateKeyUnavailable();
    }

    // we check the metadata of a random share we have on the latest polynomial we know that reflects the cloud
    // below we cater for if we have an existing share or need to create the share in the SDK
    let randomShareStore;
    const latestPolyIDOnCloud = this.lastFetchedCloudMetadata.getLatestPublicPolynomial().getPolynomialID();
    const shareIndexesExistInSDK = Object.keys(this.shares[latestPolyIDOnCloud]);
    const randomIndex = shareIndexesExistInSDK[Math.floor(Math.random() * (shareIndexesExistInSDK.length - 1))];
    if (shareIndexesExistInSDK.length >= 1) {
      randomShareStore = this.shares[latestPolyIDOnCloud][randomIndex];
    } else {
      randomShareStore = this.outputShareStore(randomIndex, latestPolyIDOnCloud);
    }
    const latestRes = await this.catchupToLatestShare({
      shareStore: randomShareStore
    });
    const latestMetadata = latestRes.shareMetadata;

    // read errors for what each means
    if (latestMetadata.nonce > this.lastFetchedCloudMetadata.nonce) {
      throw errors.acquireLockFailed(`unable to acquire write access for metadata due to 
      lastFetchedCloudMetadata (${this.lastFetchedCloudMetadata.nonce})
           being lower than last written metadata nonce (${latestMetadata.nonce}). perhaps update metadata SDK (create new tKey and init)`);
    } else if (latestMetadata.nonce < this.lastFetchedCloudMetadata.nonce) {
      throw errors.acquireLockFailed(`unable to acquire write access for metadata due to 
      lastFetchedCloudMetadata (${this.lastFetchedCloudMetadata.nonce})
      being higher than last written metadata nonce (${latestMetadata.nonce}). this should never happen as it 
      should only ever be updated by getting metadata)`);
    }
    const res = await this.storageLayer.acquireWriteLock({
      privKey: this.privKey
    });
    if (res.status !== 1) throw errors.acquireLockFailed(`lock cannot be acquired from storage layer status code: ${res.status}`);

    // increment metadata nonce for write session
    // this.metadata.nonce += 1;
    this.haveWriteMetadataLock = res.id;
    return this.metadata.nonce;
  }
  async releaseWriteMetadataLock() {
    if (!this.haveWriteMetadataLock) throw errors.releaseLockFailed("releaseWriteMetadataLock - don't have metadata lock to release");
    const res = await this.storageLayer.releaseWriteLock({
      privKey: this.privKey,
      id: this.haveWriteMetadataLock
    });
    if (res.status !== 1) throw errors.releaseLockFailed(`lock cannot be released from storage layer status code: ${res.status}`);
    this.haveWriteMetadataLock = "";
  }

  // Module functions

  async _syncShareMetadata(adjustScopedStore) {
    if (!this.metadata) {
      throw errors.metadataUndefined();
    }
    const shareArray = this.getAllShareStoresForLatestPolynomial().map(x => x.share.share);
    await this.syncMultipleShareMetadata(shareArray, adjustScopedStore);
  }
  async syncMultipleShareMetadata(shares, adjustScopedStore) {
    if (!this.metadata) {
      throw errors.metadataUndefined();
    }
    this.metadata.nonce += 1;
    const newMetadataPromise = shares.map(async share => {
      const newMetadata = this.metadata.clone();
      let specificShareMetadata;
      try {
        specificShareMetadata = await this.getAuthMetadata({
          privKey: share,
          includeLocalMetadataTransitions: true
        });
      } catch (err) {
        const prettyError = await commonTypes.prettyPrintError(err);
        throw errors.authMetadataGetUnavailable(prettyError.message);
      }
      let scopedStoreToBeSet;
      if (adjustScopedStore) {
        scopedStoreToBeSet = adjustScopedStore(specificShareMetadata.scopedStore);
      } else {
        scopedStoreToBeSet = specificShareMetadata.scopedStore;
      }
      newMetadata.scopedStore = scopedStoreToBeSet;
      return newMetadata;
    });
    const newMetadata = await Promise.all(newMetadataPromise);
    return this.setAuthMetadataBulk({
      input: newMetadata,
      privKey: shares
    });
  }
  _addRefreshMiddleware(moduleName, middleware) {
    this._refreshMiddleware[moduleName] = middleware;
  }
  _addReconstructKeyMiddleware(moduleName, middleware) {
    this._reconstructKeyMiddleware[moduleName] = middleware;
  }
  _addShareSerializationMiddleware(serialize, deserialize) {
    this._shareSerializationMiddleware = {
      serialize,
      deserialize
    };
  }
  _setDeviceStorage(storeDeviceStorage) {
    if (this.storeDeviceShare) {
      throw errors.default("storeDeviceShare already set");
    }
    this.storeDeviceShare = storeDeviceStorage;
  }
  async addShareDescription(shareIndex, description, updateMetadata) {
    if (!this.metadata) {
      throw errors.metadataUndefined();
    }
    this.metadata.addShareDescription(shareIndex, description);
    if (updateMetadata) {
      await this._syncShareMetadata();
    }
  }
  async deleteShareDescription(shareIndex, description, updateMetadata) {
    if (!this.metadata) {
      throw errors.metadataUndefined();
    }
    this.metadata.deleteShareDescription(shareIndex, description);
    if (updateMetadata) {
      await this._syncShareMetadata();
    }
  }
  async updateShareDescription(shareIndex, oldDescription, newDescription, updateMetadata) {
    if (!this.metadata) {
      throw errors.metadataUndefined();
    }
    this.metadata.updateShareDescription(shareIndex, oldDescription, newDescription);
    if (updateMetadata) {
      await this._syncShareMetadata();
    }
  }
  async encrypt(data) {
    if (!this.privKey) throw errors.privateKeyUnavailable();
    return commonTypes.encrypt(commonTypes.getPubKeyECC(this.privKey), data);
  }
  async decrypt(encryptedMessage) {
    if (!this.privKey) throw errors.privateKeyUnavailable();
    return commonTypes.decrypt(commonTypes.toPrivKeyECC(this.privKey), encryptedMessage);
  }
  async _setTKeyStoreItem(moduleName, data) {
    if (!this.metadata) {
      throw errors.metadataUndefined();
    }
    const rawTkeyStoreItems = this.metadata.getTkeyStoreDomain(moduleName) || [];
    const decryptedItems = await Promise.all(rawTkeyStoreItems.map(async x => {
      const decryptedItem = await this.decrypt(x);
      return JSON.parse(decryptedItem.toString());
    }));
    const encryptedData = await this.encrypt(Buffer.from(stringify(data)));
    const duplicateItemIndex = decryptedItems.findIndex(x => x.id === data.id);
    if (duplicateItemIndex > -1) {
      rawTkeyStoreItems[duplicateItemIndex] = encryptedData;
    } else {
      rawTkeyStoreItems.push(encryptedData);
    }

    // update metadataStore
    this.metadata.setTkeyStoreDomain(moduleName, rawTkeyStoreItems);
    await this._syncShareMetadata();
  }
  async _deleteTKeyStoreItem(moduleName, id) {
    if (!this.metadata) {
      throw errors.metadataUndefined();
    }
    const rawTkeyStoreItems = this.metadata.getTkeyStoreDomain(moduleName) || [];
    const decryptedItems = await Promise.all(rawTkeyStoreItems.map(async x => {
      const decryptedItem = await this.decrypt(x);
      return JSON.parse(decryptedItem.toString());
    }));
    const finalItems = decryptedItems.filter(x => x.id !== id);
    this.metadata.setTkeyStoreDomain(moduleName, finalItems);
    await this._syncShareMetadata();
  }
  async getTKeyStore(moduleName) {
    if (!this.metadata) {
      throw errors.metadataUndefined();
    }
    const rawTkeyStoreItems = this.metadata.getTkeyStoreDomain(moduleName) || [];
    const decryptedItems = await Promise.all(rawTkeyStoreItems.map(async x => {
      const decryptedItem = await this.decrypt(x);
      return JSON.parse(decryptedItem.toString());
    }));
    return decryptedItems;
  }
  async getTKeyStoreItem(moduleName, id) {
    if (!this.metadata) {
      throw errors.metadataUndefined();
    }
    const rawTkeyStoreItems = this.metadata.getTkeyStoreDomain(moduleName) || [];
    const decryptedItems = await Promise.all(rawTkeyStoreItems.map(async x => {
      const decryptedItem = await this.decrypt(x);
      return JSON.parse(decryptedItem.toString());
    }));
    const item = decryptedItems.find(x => x.id === id);
    return item;
  }

  // Import export shares
  async outputShare(shareIndex, type) {
    const {
      share
    } = this.outputShareStore(shareIndex).share;
    if (!type) return share;
    return this._shareSerializationMiddleware.serialize(share, type);
  }
  async inputShare(share, type) {
    if (!this.metadata) {
      throw errors.metadataUndefined();
    }
    let shareStore;
    if (!type) shareStore = this.metadata.shareToShareStore(share);else {
      const deserialized = await this._shareSerializationMiddleware.deserialize(share, type);
      shareStore = this.metadata.shareToShareStore(deserialized);
    }
    const pubPoly = this.metadata.getLatestPublicPolynomial();
    const pubPolyID = pubPoly.getPolynomialID();
    const fullShareIndexesList = this.metadata.getShareIndexesForPolynomial(pubPolyID);
    if (!fullShareIndexesList.includes(shareStore.share.shareIndex.toString("hex"))) {
      throw errors.default("Latest poly doesn't include this share");
    }
    await this.inputShareStoreSafe(shareStore);
  }
  toJSON() {
    return {
      shares: this.shares,
      enableLogging: this.enableLogging,
      privKey: this.privKey ? this.privKey.toString("hex") : undefined,
      metadata: this.metadata,
      lastFetchedCloudMetadata: this.lastFetchedCloudMetadata,
      _localMetadataTransitions: this._localMetadataTransitions,
      manualSync: this.manualSync,
      serviceProvider: this.serviceProvider,
      storageLayer: this.storageLayer
    };
  }
  getAllShareStoresForLatestPolynomial() {
    const pubPoly = this.metadata.getLatestPublicPolynomial();
    const pubPolyID = pubPoly.getPolynomialID();
    const existingShareIndexes = this.metadata.getShareIndexesForPolynomial(pubPolyID);
    const threshold = pubPoly.getThreshold();
    const pointsArr = [];
    const sharesForExistingPoly = Object.keys(this.shares[pubPolyID]);
    if (sharesForExistingPoly.length < threshold) {
      throw errors.unableToReconstruct("not enough shares for polynomial reconstruction");
    }
    for (let i = 0; i < threshold; i += 1) {
      pointsArr.push(new commonTypes.Point(new BN(sharesForExistingPoly[i], "hex"), this.shares[pubPolyID][sharesForExistingPoly[i]].share.share));
    }
    const currentPoly = lagrangeInterpolatePolynomial.lagrangeInterpolatePolynomial(pointsArr);
    const allExistingShares = currentPoly.generateShares(existingShareIndexes);
    const shareArray = existingShareIndexes.map(shareIndex => {
      return this.metadata.shareToShareStore(allExistingShares[shareIndex].share);
    });
    return shareArray;
  }

  /// Destructive method. All data will be wiped!
  async CRITICAL_deleteTkey() {
    if (!this.metadata) {
      throw errors.metadataUndefined();
    }
    if (!this.privKey) {
      throw errors.privateKeyUnavailable();
    }
    if (this._localMetadataTransitions[0].length > 0 || this._localMetadataTransitions[1].length > 0) {
      throw errors.default("Please sync all local state before calling this function");
    }

    // Construct all shares
    const shareArray = this.getAllShareStoresForLatestPolynomial();
    await this.addLocalMetadataTransitions({
      input: [...Array(shareArray.length).fill({
        message: commonTypes.SHARE_DELETED,
        dateAdded: Date.now()
      }), {
        message: commonTypes.KEY_NOT_FOUND
      }],
      privKey: [...shareArray.map(x => x.share.share), undefined]
    });
    await this.syncLocalMetadataTransitions(); // forcesync

    this.privKey = undefined;
    this.metadata = undefined;
    this.shares = {};
    this.lastFetchedCloudMetadata = undefined;
  }
  getApi() {
    return {
      getMetadata: this.getMetadata.bind(this),
      getStorageLayer: this.getStorageLayer.bind(this),
      initialize: this.initialize.bind(this),
      catchupToLatestShare: this.catchupToLatestShare.bind(this),
      _syncShareMetadata: this._syncShareMetadata.bind(this),
      _addRefreshMiddleware: this._addRefreshMiddleware.bind(this),
      _addReconstructKeyMiddleware: this._addReconstructKeyMiddleware.bind(this),
      _addShareSerializationMiddleware: this._addShareSerializationMiddleware.bind(this),
      addShareDescription: this.addShareDescription.bind(this),
      generateNewShare: this.generateNewShare.bind(this),
      inputShareStore: this.inputShareStore.bind(this),
      inputShareStoreSafe: this.inputShareStoreSafe.bind(this),
      outputShareStore: this.outputShareStore.bind(this),
      inputShare: this.inputShare.bind(this),
      outputShare: this.outputShare.bind(this),
      _setDeviceStorage: this._setDeviceStorage.bind(this),
      encrypt: this.encrypt.bind(this),
      decrypt: this.decrypt.bind(this),
      getTKeyStore: this.getTKeyStore.bind(this),
      getTKeyStoreItem: this.getTKeyStoreItem.bind(this),
      _setTKeyStoreItem: this._setTKeyStoreItem.bind(this),
      _deleteTKeyStoreItem: this._deleteTKeyStoreItem.bind(this),
      deleteShare: this.deleteShare.bind(this)
    };
  }
  setModuleReferences() {
    Object.keys(this.modules).map(x => this.modules[x].setModuleReferences(this.getApi()));
  }
  async initializeModules() {
    return Promise.all(Object.keys(this.modules).map(x => this.modules[x].initialize()));
  }
  async _refreshShares(threshold, newShareIndexes, previousPolyID) {
    if (!this.metadata) {
      throw errors.metadataUndefined();
    }
    if (!this.privKey) {
      throw errors.privateKeyUnavailable();
    }
    if (threshold > newShareIndexes.length) {
      throw errors.default(`threshold should not be greater than share indexes. ${threshold} > ${newShareIndexes.length}`);
    }

    // update metadata nonce
    this.metadata.nonce += 1;
    const poly = lagrangeInterpolatePolynomial.generateRandomPolynomial(threshold - 1, this.privKey);
    const shares = poly.generateShares(newShareIndexes);
    const existingShareIndexes = this.metadata.getShareIndexesForPolynomial(previousPolyID);
    const pointsArr = [];
    const sharesForExistingPoly = Object.keys(this.shares[previousPolyID]);
    if (sharesForExistingPoly.length < threshold) {
      throw errors.unableToReconstruct("not enough shares for polynomial reconstruction");
    }
    for (let i = 0; i < threshold; i += 1) {
      pointsArr.push(new commonTypes.Point(new BN(sharesForExistingPoly[i], "hex"), this.shares[previousPolyID][sharesForExistingPoly[i]].share.share));
    }
    const oldPoly = lagrangeInterpolatePolynomial.lagrangeInterpolatePolynomial(pointsArr);
    const shareIndexesNeedingEncryption = [];
    for (let index = 0; index < existingShareIndexes.length; index += 1) {
      const shareIndexHex = existingShareIndexes[index];
      // define shares that need encryption/relaying
      if (newShareIndexes.includes(shareIndexHex)) {
        shareIndexesNeedingEncryption.push(shareIndexHex);
      }
    }

    // add metadata new poly to metadata
    this.metadata.addFromPolynomialAndShares(poly, shares);

    // change to share stores for public storing
    const oldShareStores = {};
    const newShareStores = {};
    const polyID = poly.getPolynomialID();
    newShareIndexes.forEach(shareIndexHex => {
      newShareStores[shareIndexHex] = new commonTypes.ShareStore(shares[shareIndexHex], polyID);
    });

    // evaluate oldPoly for old shares and set new metadata with encrypted share for new polynomial

    const m = this.metadata.clone();
    const newScopedStore = {};
    const sharesToPush = await Promise.all(shareIndexesNeedingEncryption.map(async shareIndex => {
      const oldShare = oldPoly.polyEval(new BN(shareIndex, "hex"));
      const encryptedShare = await commonTypes.encrypt(commonTypes.getPubKeyECC(oldShare), Buffer.from(JSON.stringify(newShareStores[shareIndex])));
      newScopedStore[commonTypes.getPubKeyPoint(oldShare).x.toString("hex")] = encryptedShare;
      oldShareStores[shareIndex] = new commonTypes.ShareStore(new commonTypes.Share(shareIndex, oldShare), previousPolyID);
      return oldShare;
    }));
    m.setScopedStore("encryptedShares", newScopedStore);
    const metadataToPush = Array(sharesToPush.length).fill(m);

    // run refreshShare middleware
    // If a shareIndex is left out during refresh shares, we assume that it being explicitly deleted.
    for (const moduleName in this._refreshMiddleware) {
      if (Object.prototype.hasOwnProperty.call(this._refreshMiddleware, moduleName)) {
        const adjustedGeneralStore = this._refreshMiddleware[moduleName](this.metadata.getGeneralStoreDomain(moduleName), oldShareStores, newShareStores);
        if (!adjustedGeneralStore) this.metadata.deleteGeneralStoreDomain(moduleName);else this.metadata.setGeneralStoreDomain(moduleName, adjustedGeneralStore);
      }
    }
    const newShareMetadataToPush = [];
    const newShareStoreSharesToPush = newShareIndexes.map(shareIndex => {
      const me = this.metadata.clone();
      newShareMetadataToPush.push(me);
      return newShareStores[shareIndex].share.share;
    });
    const AuthMetadatas = this.generateAuthMetadata({
      input: [...metadataToPush, ...newShareMetadataToPush]
    });

    // Combine Authmetadata and service provider ShareStore
    await this.addLocalMetadataTransitions({
      input: [...AuthMetadatas, newShareStores["1"]],
      privKey: [...sharesToPush, ...newShareStoreSharesToPush, undefined]
    });

    // update this.shares with these new shares
    for (let index = 0; index < newShareIndexes.length; index += 1) {
      const shareIndex = newShareIndexes[index];
      this.inputShareStore(newShareStores[shareIndex]);
    }
    // await this.releaseWriteMetadataLock();
    return {
      shareStores: newShareStores
    };
  }
  async _initializeNewKey({
    determinedShare,
    initializeModules,
    importedKey,
    importEd25519Seed,
    delete1OutOf1
  } = {}) {
    if (!importedKey) {
      const tmpPriv = lagrangeInterpolatePolynomial.generatePrivateBN();
      this.secp256k1Key = tmpPriv;
    } else {
      this.secp256k1Key = importedKey;
    }

    // create a random poly and respective shares
    // 1 is defined as the serviceProvider share
    // 0 is for tKey
    const shareIndexForDeviceStorage = commonTypes.generatePrivateExcludingIndexes([new BN(1), new BN(0)]);
    const shareIndexes = [new BN(1), shareIndexForDeviceStorage];
    let poly;
    if (determinedShare) {
      const shareIndexForDeterminedShare = commonTypes.generatePrivateExcludingIndexes([new BN(1), new BN(0)]);
      poly = lagrangeInterpolatePolynomial.generateRandomPolynomial(1, this.privKey, [new commonTypes.Share(shareIndexForDeterminedShare, determinedShare)]);
      shareIndexes.push(shareIndexForDeterminedShare);
    } else {
      poly = lagrangeInterpolatePolynomial.generateRandomPolynomial(1, this.privKey);
    }
    const shares = poly.generateShares(shareIndexes);

    // create metadata to be stored
    const metadata$1 = new metadata(commonTypes.getPubKeyPoint(this.privKey));
    metadata$1.addFromPolynomialAndShares(poly, shares);
    const serviceProviderShare = shares[shareIndexes[0].toString("hex")];
    const shareStore = new commonTypes.ShareStore(serviceProviderShare, poly.getPolynomialID());
    this.metadata = metadata$1;

    // setup ed25519 seed after metadata is set
    // import/gen ed25519 seed
    await this.setupEd25519Seed(importEd25519Seed);

    // initialize modules
    if (initializeModules) {
      await this.initializeModules();
    }
    const metadataToPush = [];
    const sharesToPush = shareIndexes.map(shareIndex => {
      metadataToPush.push(this.metadata);
      return shares[shareIndex.toString("hex")].share;
    });
    const authMetadatas = this.generateAuthMetadata({
      input: metadataToPush
    });

    // because this is the first time we're setting metadata there is no need to acquire a lock
    // acquireLock: false. Force push
    await this.addLocalMetadataTransitions({
      input: [...authMetadatas, shareStore],
      privKey: [...sharesToPush, undefined]
    });
    if (delete1OutOf1) {
      await this.addLocalMetadataTransitions({
        input: [{
          message: commonTypes.ONE_KEY_DELETE_NONCE
        }],
        privKey: [this.serviceProvider.postboxKey]
      });
    }

    // store metadata on metadata respective to shares
    for (let index = 0; index < shareIndexes.length; index += 1) {
      const shareIndex = shareIndexes[index];
      // also add into our share store
      this.inputShareStore(new commonTypes.ShareStore(shares[shareIndex.toString("hex")], poly.getPolynomialID()));
    }
    if (this.storeDeviceShare) {
      await this.storeDeviceShare(new commonTypes.ShareStore(shares[shareIndexes[1].toString("hex")], poly.getPolynomialID()));
    }
    const result = {
      secp256k1Key: this.privKey,
      deviceShare: new commonTypes.ShareStore(shares[shareIndexes[1].toString("hex")], poly.getPolynomialID()),
      userShare: undefined
    };
    if (determinedShare) {
      result.userShare = new commonTypes.ShareStore(shares[shareIndexes[2].toString("hex")], poly.getPolynomialID());
    }
    return result;
  }
  async importEd25519Seed(seed) {
    if (!this.privKey) {
      throw errors.privateKeyUnavailable();
    }
    if (this.getEd25519PublicKey()) {
      throw errors.default("Ed25519 key already exists");
    }

    // derive key pair (scalar, public key point) from seed
    const keyPair = torus_js.getEd25519ExtendedPublicKey(seed);
    this.metadata.setGeneralStoreDomain(ed25519SeedConst, {
      message: await this.encrypt(seed),
      publicKey: keyPair.point.encode("hex", false)
    });
    this._ed25519Seed = seed;
  }
  async setupEd25519Seed(seed) {
    if (!this.privKey) {
      throw errors.privateKeyUnavailable();
    }
    let seedToUse = seed;
    if (!seed) {
      const newEd25519Seed = await random.getRandomBytes(32);
      seedToUse = Buffer.from(newEd25519Seed);
    }
    await this.importEd25519Seed(seedToUse);
  }
}
var ThresholdKey$1 = ThresholdKey;

module.exports = ThresholdKey$1;
