// @tkey/common-types/dist/lib.cjs/index.js
'use strict';

var commonTypes = require('./baseTypes/commonTypes.js');
var utils = require('./utils.js');
var Point = require('./base/Point.js');
var Polynomial = require('./base/Polynomial.js');
var PublicPolynomial = require('./base/PublicPolynomial.js');
var PublicShare = require('./base/PublicShare.js');
var Share = require('./base/Share.js');
var ShareStore = require('./base/ShareStore.js');
var BNUtils = require('./base/BNUtils.js');
var Error = require('./base/Error.js');
var OneKey = require('./base/OneKey.js');



exports.KeyType = commonTypes.KeyType;
exports.KEY_NOT_FOUND = utils.KEY_NOT_FOUND;
exports.SHARE_DELETED = utils.SHARE_DELETED;
exports.decrypt = utils.decrypt;
exports.derivePubKeyXFromPolyID = utils.derivePubKeyXFromPolyID;
exports.encrypt = utils.encrypt;
exports.generateAddressFromPublicKey = utils.generateAddressFromPublicKey;
exports.generateID = utils.generateID;
exports.generatePrivateExcludingIndexes = utils.generatePrivateExcludingIndexes;
exports.isEmptyObject = utils.isEmptyObject;
exports.isErrorObj = utils.isErrorObj;
exports.normalize = utils.normalize;
exports.prettyPrintError = utils.prettyPrintError;
exports.secp256k1 = utils.secp256k1;
exports.stripHexPrefix = utils.stripHexPrefix;
exports.Point = Point;
exports.Polynomial = Polynomial;
exports.PublicPolynomial = PublicPolynomial;
exports.PublicShare = PublicShare;
exports.Share = Share;
exports.ShareStore = ShareStore;
exports.getPubKeyEC = BNUtils.getPubKeyEC;
exports.getPubKeyECC = BNUtils.getPubKeyECC;
exports.getPubKeyPoint = BNUtils.getPubKeyPoint;
exports.toPrivKeyEC = BNUtils.toPrivKeyEC;
exports.toPrivKeyECC = BNUtils.toPrivKeyECC;
exports.TkeyError = Error.TkeyError;
exports.ONE_KEY_DELETE_NONCE = OneKey.ONE_KEY_DELETE_NONCE;
exports.ONE_KEY_NAMESPACE = OneKey.ONE_KEY_NAMESPACE;
