// @tkey/share-serialization/dist/lib.cjs/errors.js
'use strict';

var _defineProperty = require('@babel/runtime/helpers/defineProperty');
var commonTypes = require('@tkey/common-types');

class ShareSerializationError extends commonTypes.TkeyError {
  constructor(code, message) {
    // takes care of stack and proto
    super(code, message);
    // Set name explicitly as minification can mangle class names
    Object.defineProperty(this, "name", {
      value: "ShareSerializationError"
    });
  }
  static fromCode(code, extraMessage = "") {
    return new ShareSerializationError(code, `${ShareSerializationError.messages[code]}${extraMessage}`);
  }
  static default(extraMessage = "") {
    return new ShareSerializationError(7000, `${ShareSerializationError.messages[7000]}${extraMessage}`);
  }

  // Custom methods
  static typeNotSupported(extraMessage = "") {
    return ShareSerializationError.fromCode(7010, extraMessage);
  }
  static invalidEntropy(extraMessage = "") {
    return ShareSerializationError.fromCode(7011, extraMessage);
  }
  static invalidChecksum(extraMessage = "") {
    return ShareSerializationError.fromCode(7012, extraMessage);
  }
  static invalidMnemonic(extraMessage = "") {
    return ShareSerializationError.fromCode(7013, extraMessage);
  }
}
_defineProperty(ShareSerializationError, "messages", {
  7000: "Custom",
  // Misc
  7010: "Type is not supported",
  7011: "Invalid Entropy",
  7012: "Invalid Checksum",
  7013: "Invalid mnemonic"
});
var ShareSerializationError$1 = ShareSerializationError;

module.exports = ShareSerializationError$1;
