// @tkey/share-serialization/dist/lib.esm/ShareSerializationModule.js
import _defineProperty from '@babel/runtime/helpers/defineProperty';
import BN from 'bn.js';
import { english } from './english.js';
import ShareSerializationError from './errors.js';
import { entropyToMnemonic, mnemonicToEntropy } from './utils.js';

const SHARE_SERIALIZATION_MODULE_NAME = "shareSerialization";
class ShareSerializationModule {
  constructor() {
    _defineProperty(this, "moduleName", void 0);
    _defineProperty(this, "tbSDK", void 0);
    this.moduleName = SHARE_SERIALIZATION_MODULE_NAME;
  }
  static serializeMnemonic(share) {
    return entropyToMnemonic(share.toString("hex").padStart(64, "0"), english);
  }
  static deserializeMnemonic(share) {
    return new BN(mnemonicToEntropy(share, english), "hex");
  }
  setModuleReferences(tbSDK) {
    this.tbSDK = tbSDK;
    this.tbSDK._addShareSerializationMiddleware(this.serialize.bind(this), this.deserialize.bind(this));
  }

  // eslint-disable-next-line
  async initialize() {}
  async serialize(share, type) {
    if (type === "mnemonic") {
      return ShareSerializationModule.serializeMnemonic(share);
    }
    throw ShareSerializationError.typeNotSupported();
  }
  async deserialize(serializedShare, type) {
    if (type === "mnemonic") return ShareSerializationModule.deserializeMnemonic(serializedShare);
    throw ShareSerializationError.typeNotSupported();
  }
}
var ShareSerializationModule$1 = ShareSerializationModule;

export { SHARE_SERIALIZATION_MODULE_NAME, ShareSerializationModule$1 as default };
